docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = (\arg2 -> do
+  let
+    a = { bar: 41 }
+  1 + a.bar + 8 + arg2
+  ) 9
+```
+
+At this point, `purs-backend-es`' default inliners will see a lambda (i.e. a "producer") being immediately applied to an argument (i.e. a "consumer"). Thus, it will inline that argument into the lambda's body. This gets us:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + 8 + 9
+```
+
+A rewrite rule will detect the `8 + 9` expression, and replace it with its sum: `17`. However, `1 + a.bar` and `a.bar + 17` don't trigger the rewrite rule because one of the values is not a literal `Int` value.
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + 17
+```
+
+Fortunately, `purs-backend-es`' default inliners will see a record binding (i.e. a "producer") that is immediately accessed under the label `bar`. Thus, the corresponding value will be inlined:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  1 + 41 + 17
+```
+
+The rewrite rules then detects that `1 + 41` is `42`:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  42 + 17
+```
+
+The rewrite rules then detects that `42 + 17` is `59`:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  59
+```
+
+Lastly, `purs-backend-es` determines that the `a` binding is never used; thus, it's removed entirely:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = 59
+```
+
+Now that we've finished optimizing `foo`, the following states are true:
+
+- `ignoreArgs` is still defined as it originally was
+- the optimized `foo` still represents the same value as the unoptimized `foo` would have had at runtime.
+
+## A Methodology for Defining Inline Directives
+
+Inlining duplicates code so that rewrite rules can trigger optimizations that were otherwise hidden from its eyes. Ideally, inlining values will always trigger rewrite rules that both reduce the bundle size of the code AND make the resulting program more performant. However, inlining may increase a program's bundle size without improving its performance.
+
+In short, this process isn't scientific. While one can add inline directives without much thought, the result likely won't be what they want.
+
+So, how should one determine if an inline directive needs to be added? One should always use the below methodology:
+
+1. Think of a snippet of code you want to ensure is optimized.
+2. Define a snapshot for that snippet.
```

Comment:
What is a snapshot and how does the reader define one? There's a lot of implicit knowledge here that makes no sense to most readers.

Reply:
Is the audience contributors to the project, or users of the backend?

Reply:
The intended audience are users of the backend. I guess rather than 'snapshot', I should continue using 'snippet'. However, an example that clarifies what a good snippet would look like may help illustrate/define this term more.